tree.go

```diff
@@ -1293,3 +1266,155 @@ func ToDot(root VerkleNode) string {
+			count := 0
+			for _, c := range n.children {
+				_, isInternalNode := c.(*InternalNode)
+				_, isLeafNode := c.(*LeafNode)
+				if isInternalNode || isLeafNode {
+					pointsToCompress = append(pointsToCompress, c.Commitment())
+					count++
+				}
+			}
+			pointToCompressPerNodeCount[i] = 1 + count // Internal node commitment + all non-hashed children.
+		case *LeafNode:
+			pointsToCompress = append(pointsToCompress, n.commitment, n.c1, n.c2)
+			pointToCompressPerNodeCount[i] = 3 // Leaf node commitment + c1 + c2.
+		}
+	}
+
+	// Now we do the all transformations in a single-shot.
+	compressedPoints := banderwagon.ElementsToBytes(pointsToCompress)
+
+	// Now we that we did the heavy CPU work, we have to do the rest of `nodes` serialization
+	// taking the compressed points from this single list.
+	ret := make([]SerializedNode, 0, len(nodes))
+	for i := range nodes {
+		switch n := nodes[i].(type) {
+		case *InternalNode:
+			compressedChildren := compressedPoints[1:pointToCompressPerNodeCount[i]]
+			sn := SerializedNode{
+				Node:            n,
+				CommitmentBytes: compressedPoints[0],
+				SerializedBytes: n.serializeWithCompressedChildren(compressedChildren),
+			}
+			ret = append(ret, sn)
+		case *LeafNode:
+			c1Bytes := compressedPoints[0]
+			c2Bytes := compressedPoints[1]
+			sn := SerializedNode{
+				Node:            n,
+				CommitmentBytes: compressedPoints[0],
+				SerializedBytes: n.serializeWithCompressedCommitments(c1Bytes, c2Bytes),
+			}
+			ret = append(ret, sn)
+		}
+		compressedPoints = compressedPoints[pointToCompressPerNodeCount[i]:]
+	}
+
+	return ret, nil
+}
+
+func (n *InternalNode) collectNonHashedNodes(list []VerkleNode) []VerkleNode {
+	list = append(list, n)
+	for _, child := range n.children {
+		switch childNode := child.(type) {
+		case *LeafNode:
+			list = append(list, childNode)
+		case *InternalNode:
+			list = childNode.collectNonHashedNodes(list)
+		}
+	}
+	return list
+}
+
+func (n *InternalNode) serializeWithCompressedChildren(compressedChildren [][32]byte) []byte {
+	var (
+		bitlist, hashlist [NodeWidth / 8]byte
+		nhashed           int // number of children who are hashed nodes
+	)
+	for i, c := range n.children {
+		if _, ok := c.(Empty); !ok {
+			setBit(bitlist[:], i)
+			if _, ok := c.(*HashedNode); ok {
+				setBit(hashlist[:], i)
+				nhashed++
+			}
+		}
+	}
+
+	ret := make([]byte, nodeTypeSize+bitlistSize+(len(compressedChildren)+nhashed)*SerializedPointCompressedSize)
+
+	children := ret[internalNodeChildrenOffset:internalNodeChildrenOffset]
+	consumed := 0
+	for i := 0; i < NodeWidth; i++ {
+		if bit(bitlist[:], i) {
+			if bit(hashlist[:], i) {
+				children = append(children, n.children[i].(*HashedNode).commitment...)
+			} else {
+				children = append(children, compressedChildren[consumed][:]...)
+				consumed++
+			}
+		}
+	}
+
+	// Store in ret the serialized result
+	ret[nodeTypeOffset] = internalRLPType
+	copy(ret[internalBitlistOffset:], bitlist[:])
```

Comment:
why not use `ret[internalBitlistOffset:]` directly instead of copying the bitlist? Exactly what you do with `children` and `ret`.

Reply:
Because `bitlist` is used and filled before `ret` is defined. 

To define `ret` here in L1369 we use/need `nhashed` (to calculate the length). And we calculate `nhashed` when constructing `bitlist` (L1361).

What we could do is move `ret` to line ~L1358, and maybe assume a `len(compressedChildren)+nhashed` to be a worst case scenario, so we don't need the exact `nhashed`. After we have `nhashed` we can redefine `ret` to have this exact length again, and do the same.

The only tradeoff is reserving some extra unneeded capacity, but that can avoid the `copy` here (which is a 32 byte copy, so not sure if will make a difference but we can try).

Reply:
OK, implemented the idea to be able to do this. 32 bytes isn't that bad to copy, but also some extra bytes in memory shouldn't hurt.

Done!